pycbc/inference/sampler/refine.py

```diff
@@ -0,0 +1,234 @@
+""" Sampler that uses kde refinement of an existing posterior estimate.
+"""
+
+import logging
+import numpy
+import numpy.random
+
+from scipy.special import logsumexp
+from scipy.stats import gaussian_kde
+from scipy.stats import entropy as sentropy
+
+from pycbc.inference import models
+from pycbc.pool import choose_pool
+from pycbc.inference.io import loadfile
+
+from .base import setup_output, initial_dist_from_config
+from .dummy import DummySampler
+
+
+def call_model(params):
+    models._global_instance.update(**params)
+    return (models._global_instance.logposterior,
+            models._global_instance.loglikelihood)
+
+
+def resample_equal(samples, logwt, seed=0):
+    weights = numpy.exp(logwt - logsumexp(logwt))
+    N = len(weights)
+    positions = (numpy.random.random() + numpy.arange(N)) / N
+    idx = numpy.zeros(N, dtype=int)
+    cumulative_sum = numpy.cumsum(weights)
+    cumulative_sum /= cumulative_sum[-1]
+    i, j = 0, 0
+    while i < N:
+        if positions[i] < cumulative_sum[j]:
+            idx[i] = j
+            i += 1
+        else:
+            j += 1
+    try:
+        rng = numpy.random.default_rng(seed)
+    except AttributeError:
+        # numpy pre-1.17 uses RandomState
+        # Py27: delete this after we drop python 2.7 support
+        rng = numpy.random.RandomState(seed)
+    rng.shuffle(idx)
+    return {p: samples[p][idx] for p in samples}
+
+
+class RefineSampler(DummySampler):
+    """Sampler for kde drawn refinement of existing posterior estimate
+
```

Comment:
Can we provide an example of how this might work in the code/documentation? I assume we generate a normal output, and then run this. Is it valid for both nested sampling and MCMC?